File: ruby/lib/base_to_dec.rb
# frozen_string_literal: true

# BaseToDec :
class BaseToDec
  CHARSET = '0123456789ABCDEF'
  def self.iterate(value, base)
    res = 0
    multiplier = 1
    (value.length - 1).downto(0).each do |n|
      index = CHARSET.index(value[n])
      raise 'something went wrong.' if index.nil?

      res += index * multiplier
      multiplier *= base
    end
    res
  end

  def self.std(value, base)
    value.to_i(base)
  end
end

File: rust/src/base_to_dec.rs
pub fn iterate(value: String, base: u32) -> u32 {
    const CHARSET: &str = "0123456789ABCDEF";
    let mut res = 0 as u32;
    let mut multiplier = 1 as u64;
    for c in value.chars().rev() {
        let index = match CHARSET.find(c) {
            Some(x) => x as u32,
            None => panic!("something went wrong!"),
        };
        res += index * multiplier as u32;
        multiplier *= base as u64;
    }
    res
}

pub fn std(value: String, base: u32) -> u32 {
    match u64::from_str_radix(value.as_str(), base) {
        Ok(v) => v as u32,
        Err(_) => panic!("something went wrong!"),
    }
}
